shims/src/peschke/shims/resource/Bracket.scala

```diff
@@ -0,0 +1,110 @@
+  */
+trait Bracket[F[_]] {
+
+  /** Initialize, use, and release some resource.
+    *
+    * The resource should not be assumed to be available outside of the scope of
+    * `use`
+    */
+  def bracket[R, A]
+    (acquire: () => F[R], use: R => F[A], release: R => F[Complete])
+    : F[A]
+}
+object Bracket {
+  def apply[F[_]](implicit B: Bracket[F]): B.type = B
+
+  implicit def bracketFuture(implicit ec: ExecutionContext): Bracket[Future] =
+    new Bracket[Future] {
+      override def bracket[R, A](
+                                  acquire: () => Future[R],
+                                  use:     R => Future[A],
+                                  release: R => Future[Complete]
+      ): Future[A] =
+        acquire().flatMap { resource =>
+          use(resource).transformWith {
+            case Success(value) => release(resource).map(_ => value)
+            case Failure(useException) =>
+              release(resource)
+                .transform {
+                  case Failure(releaseException) =>
+                    useException.addSuppressed(releaseException)
+                    Failure(useException)
+                  case Success(_) => Failure(useException)
+                }
+          }
+        }
+    }
+
+  implicit val bracketTry: Bracket[Try] =
+    new Bracket[Try] {
+      override def bracket[R, A](
+                                  acquire: () => Try[R],
+                                  use:     R => Try[A],
+                                  release: R => Try[Complete]
+      ): Try[A] =
+        acquire().flatMap { resource =>
+          use(resource).transform(
+            value => release(resource).map(_ => value),
+            useException =>
+              release(resource)
+                .transform(
+                  _ => Failure(useException),
+                  releaseException => {
+                    useException.addSuppressed(releaseException)
+                    Failure(useException)
+                  }
+                )
+          )
+        }
+    }
+
+  implicit val bracketId: Bracket[Id] = new Bracket[Id] {
+    override def bracket[R, A](
+                                acquire: () => Id[R],
+                                use:     R => Id[A],
+                                release: R => Id[Complete]
+    ): Id[A] =
+      bracketTry
+        .bracket[R, A](
+          () => Try(acquire()),
+          r => Try(use(r)),
+          r => Try(release(r))
+        ).get
+  }
+
+  implicit def bracketEither[E]
+    (implicit EM: ExceptionMapper[Either[E, *], E], SE: Semigroup[E])
+    : Bracket[Either[E, *]] =
+    new Bracket[Either[E, *]] {
+      override def bracket[R, A](
+                                  acquire: () => Either[E, R],
+                                  use:     R => Either[E, A],
+                                  release: R => Either[E, Complete])
+        : Either[E, A] =
+        EM.catchNonFatalF(acquire()).flatMap { resource =>
```

Comment:
Will a for comprehension be easier to read?

**`[not-an-action]`**

Reply:
Not applicable in this case, because we need to act on both the `Left` and `Right` results. A for-comprehension would short-circuit and skip releasing the resource if `use` fails.